### PR TITLE
variable histogram bins added

### LINKED ISSUE #2379: Ideally the bins in histogram equalization is variable rather than limited to 256
https://github.com/tensorflow/addons/blob/d26e2ed5f68092aed57016a7005ce534b1be3dce/tensorflow_addons/image/color_ops.py#L36

### PR DESCRIPTION
**Pull request by whatwilliam**
# Description

Changes the equalization method in color_ops.py to take in a variable number of histogram bins.

Fixes #2379

## Type of change

- [x] Bug fix
- [x] Updated or additional documentation

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [x] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [x] This PR addresses an already submitted issue for TensorFlow Addons
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

Function has been sanity tested to be the same as the old equalize_image and past tests should still pass. This PR adds `bins` as an optional `int` parameter which defaults to `256` to ensure backwards compatibility. The expected use case of this parameter is for imaging data which has greater dynamic range than 8-bit.

### CODE REVIEW

**Comment by Ouwen** [comment]
@whatwilliam tests fail for equalize likely some small addition error 

```
=================================== FAILURES ===================================
____________________ test_equalize_with_PIL[eager_mode-cpu] ____________________

    @pytest.mark.usefixtures("maybe_run_functions_eagerly")
    def test_equalize_with_PIL():
        np.random.seed(0)
        image = np.random.randint(low=0, high=255, size=(4, 3, 3, 3), dtype=np.uint8)
        equalized = np.stack([ImageOps.equalize(Image.fromarray(i)) for i in image])
>       np.testing.assert_equal(color_ops.equalize(tf.constant(image)).numpy(), equalized)
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 108 / 108 (100%)
E       Max absolute difference: 252
E       Max relative difference: 50.4
E        x: array([[[[0, 1, 0],
E                [0, 0, 0],
E                [0, 0, 0]],...
E        y: array([[[[172,  10, 127],
E                [140,  47, 170],
E                [196, 151, 117]],...

tensorflow_addons/image/tests/color_ops_test.py:50: AssertionError
___________________ test_equalize_with_PIL[tf_function-cpu] ____________________

    @pytest.mark.usefixtures("maybe_run_functions_eagerly")
    def test_equalize_with_PIL():
        np.random.seed(0)
        image = np.random.randint(low=0, high=255, size=(4, 3, 3, 3), dtype=np.uint8)
        equalized = np.stack([ImageOps.equalize(Image.fromarray(i)) for i in image])
>       np.testing.assert_equal(color_ops.equalize(tf.constant(image)).numpy(), equalized)
E       AssertionError: 
E       Arrays are not equal
E       
E       Mismatched elements: 108 / 108 (100%)
E       Max absolute difference: 252
E       Max relative difference: 50.4
E        x: array([[[[0, 1, 0],
E                [0, 0, 0],
E                [0, 0, 0]],...
E        y: array([[[[172,  10, 127],
E                [140,  47, 170],
E                [196, 151, 117]],...
```

File: tensorflow_addons/image/color_ops.py
Comment (Ouwen) [inline tensorflow_addons/image/color_ops.py:42]: likely

File: tensorflow_addons/image/color_ops.py
Comment (bhack) [inline tensorflow_addons/image/color_ops.py:53]: For line 43 could it clip with 0 ?

File: tensorflow_addons/image/color_ops.py
Comment (whatwilliam) [reply tensorflow_addons/image/color_ops.py]: new commit has been made to catch it, all checks now pass.

File: tensorflow_addons/image/color_ops.py
Comment (whatwilliam) [reply tensorflow_addons/image/color_ops.py:53]: If I understand your question correctly, I suppose so, then all you would be left with is a Tensor of zeros. But that would also mean that the equalization in question would be in one histogram bin, which would not be the intended use case.

Comment (bhack) [reply tensorflow_addons/image/color_ops.py:53]: Yes, probably you could test also 0, 256, and not 256 bins with https://scikit-image.org/docs/dev/api/skimage.exposure.html#skimage.exposure.equalize_hist

File: tensorflow_addons/image/color_ops.py
Comment (bhack) [inline tensorflow_addons/image/color_ops.py:34]: I suppose this is not the case for one of the API that expose bins params like scikit-image that we could already use in tests here https://scikit-image.org/docs/dev/api/skimage.exposure.html#skimage.exposure.equalize_hist

Comment (whatwilliam) [reply tensorflow_addons/image/color_ops.py]: assertion has been removed

**Review by bhack — CHANGES_REQUESTED** [review]
I think that you can switch the test from `ImageOps.equalize` to skimage.exposure.equalize_hist to test with different bins values

**Comment by whatwilliam** [comment]
> I think that you can switch the test from `ImageOps.equalize` to `skimage.exposure.equalize_hist` to test with different bins values

The purpose of this PR is to add variable histogram bins to Tensorflow Addon's Image Ops.

**Comment by bhack** [comment]
> The purpose of this PR is to add variable histogram bins to Tensorflow Addon's Image Ops.

Yes but as its related test uses PIL `ImageOps.equalize` it doesn't support bins testing:

https://github.com/tensorflow/addons/blob/704c247a6d322e19cb07640aa1861c0366b43ce3/tensorflow_addons/image/tests/color_ops_test.py#L49

**Comment by whatwilliam** [comment]
> > The purpose of this PR is to add variable histogram bins to Tensorflow Addon's Image Ops.
> 
> Yes but as its related test uses PIL `ImageOps.equalize` it doesn't support bins testing:
> 
> https://github.com/tensorflow/addons/blob/704c247a6d322e19cb07640aa1861c0366b43ce3/tensorflow_addons/image/tests/color_ops_test.py#L49

That will not work because the `equalize` function on the original `tensorflow:master` doesn't return the same array as using the `skimage.exposure.equalize_hist` given the same image at 256 bins.

**Comment by whatwilliam** [comment]
> > > The purpose of this PR is to add variable histogram bins to Tensorflow Addon's Image Ops.
> > 
> > 
> > Yes but as its related test uses PIL `ImageOps.equalize` it doesn't support bins testing:
> > https://github.com/tensorflow/addons/blob/704c247a6d322e19cb07640aa1861c0366b43ce3/tensorflow_addons/image/tests/color_ops_test.py#L49
> 
> That will not work because the `equalize` function on the original `tensorflow:master` doesn't return the same array as using the `skimage.exposure.equalize_hist` given the same image at 256 bins. It also seems that skimage normalizes their result from [0,1] but even when the original equalize function from tensorflow's result is normalized, the two arrays differ by about 0.018

Since both implementations don't give the same answer, would it be alright to create a test that tests for a acceptable range of difference using skimage?

**Comment by bhack** [comment]
> Since both implementations don't give the same answer, would it be alright to create a test that tests for a acceptable range of difference using skimage?

Just if the numerical tolerance is minimal and related to differences in precision. 
Or you need to find another way to refactor the test to verify bins.

/cc @tomerk We have a duplicated impl at https://github.com/tensorflow/models/blob/master/official/vision/beta/ops/augment.py#L476-L506

File: tensorflow_addons/image/color_ops.py
Comment (WindQAQ) [inline tensorflow_addons/image/color_ops.py:31]: `bins: int = 256`

File: tensorflow_addons/image/color_ops.py
Comment (WindQAQ) [inline tensorflow_addons/image/color_ops.py:56]: `bins: int = 256`

File: tensorflow_addons/image/color_ops.py
Comment (WindQAQ) [inline tensorflow_addons/image/color_ops.py:66]: `bins: int = 256`

`Optional[int]` means `Union[None, int]`. I believe it does not work when bins is passed as `None`.

**Review by WindQAQ — CHANGES_REQUESTED** [review]

File: tensorflow_addons/image/color_ops.py
Comment (whatwilliam) [reply tensorflow_addons/image/color_ops.py]: `Optional` has been removed. It is now `bins: int = 256`

File: tensorflow_addons/image/tests/color_ops_test.py
Comment (bhack) [inline tensorflow_addons/image/tests/color_ops_test.py:67]: I didn't go into details but I'm still a little bit concerned about these numerical differences on a such small input matrix. @WindQAQ What do you think?

File: tensorflow_addons/image/tests/color_ops_test.py
Comment (bhack) [inline tensorflow_addons/image/tests/color_ops_test.py:54]: Probably we could add `@pytest.mark.parametrize` to test different bins

File: tensorflow_addons/image/tests/color_ops_test.py
Comment (whatwilliam) [reply tensorflow_addons/image/tests/color_ops_test.py:67]: The 1.5% metric comes from the difference between the implementations of the proven `PIL.ImageOps.equalize` and the test I'm using with `skimage.exposure.equalize_hist`.

Here's a quick showcase I made
```
#compares PIL and skimage
import tensorflow as tf
import numpy as np
import PIL
from skimage.exposure import equalize_hist
from PIL.ImageOps import *

np.random.seed(0)

low_image = np.random.randint(low=0, high=256, size=(4, 3, 3, 3), dtype=np.uint8)
PIL_equalized = np.stack([PIL.ImageOps.equalize(PIL.Image.fromarray(i)) for i in low_image])
PIL_equalized = tf.math.divide(tf.cast(PIL_equalized, tf.float32), 255) #normalized
skiequalized = equalize_hist(low_image, 256)
low_avg = tf.reduce_mean(
            tf.math.abs(tf.reshape(tf.math.subtract(PIL_equalized, skiequalized), [-1]))
)

high_image = np.random.randint(low=0, high=256, size=(28, 28, 3, 3), dtype=np.uint8)
PIL_equalized = np.stack([PIL.ImageOps.equalize(PIL.Image.fromarray(i)) for i in high_image])
PIL_equalized = tf.math.divide(tf.cast(PIL_equalized, tf.float32), 255) #normalized
skiequalized = equalize_hist(high_image, 256)
high_avg = tf.reduce_mean(
            tf.math.abs(tf.reshape(tf.math.subtract(PIL_equalized, skiequalized), [-1]))
)
print(low_avg, high_avg)
```
```
OUTPUT:
tf.Tensor(0.036022358, shape=(), dtype=float32) tf.Tensor(0.0061728833, shape=(), dtype=float32)
```
Here the differences between the two ranges from 3.6-0.6% depending on the amount of variance allowed by size of the image/number of pixels compared. So 1.5% is a reasonable metric that stands between this range that accounts for the different implementations

File: tensorflow_addons/image/tests/color_ops_test.py
Comment (Ouwen) [reply tensorflow_addons/image/tests/color_ops_test.py:54]: @bhack the two most common would probably be 256 and 2^16. I'm not sure if it is a good idea to include more as it would slow the tests down. Additionally, since the test matrices are randomly generated there is a chance a low error tolerance would cause intermittent failures.

I think having 1 test to match the current PIL function and another test to match skimage within some tolerance that doesn't cause transient errors would be ideal. Otherwise we may need to adopt skimage equalize implementation which may be a much larger change than the current PR.

Another way to solve the transient error issue is to load a static test image.

Comment (bhack) [reply tensorflow_addons/image/tests/color_ops_test.py:54]: Having the two most common is ok. How many bins are testing skimage?

Comment (whatwilliam) [reply tensorflow_addons/image/tests/color_ops_test.py:54]: @bhack Originally only one random bin size was tested. The most recent commit adds `@pytest.mark.parametrize` to test for bin sizes 256 and 2^16.

File: tensorflow_addons/image/tests/color_ops_test.py
Comment (bhack) [reply tensorflow_addons/image/tests/color_ops_test.py:67]: Please add a comment in the test why we need this tolerance

Comment (whatwilliam) [reply tensorflow_addons/image/tests/color_ops_test.py]: comment has been added

**Review by bhack — APPROVED** [review]
I approve but I will not merge this until I have a second approval from another reviewer for the current  tolerance approach over an alternative implementation in tests

**Review by WindQAQ — APPROVED** [review]

**Pull request merged**
